### ISSUE TITLE
Bug: Inconsistent behavior between development and production builds

### ISSUE THREAD

**Issue by dutzi** [facebook/react#20938]
I was trying to implement `useOn`, a hook lets you write the following code:

```js
useOn()
  .who(window)
  .when('resize')
  .what(() => {
    console.log('resize!');
  });
```

My approach worked in production, but not in development. I discovered some pretty unexpected behavior by React.

When running `useOn` in development, it seemed like the resize event listener isn't being removed when the ResizeComponent gets unmounted (turned out it was actually being added twice).

So I added a `console.log`, just above the call to `addEventListener`, but only one log appeared.

I then added a `debugger` statement just above the call to `console.log` and noticed that it is indeed being called twice, but in the second time it's called, `console.log` is overwritten by React's `disabledLog()`.

React version: 17.0.1

## Steps To Reproduce

(tip: you can skip checking out the repo (steps 1 to 3), as it can be reproduced in [the repo's CodeSandbox](https://codesandbox.io/s/github/dutzi/use-on))

1. Check out https://github.com/dutzi/use-on
2. `yarn` then `yarn start`
3. (Chrome will start up)
4. Open the DevTools' Console
5. Click on "Mount Resize Component"
6. Resize window
7. (You should see "resize!" printed in the console)
8. Click on "Unmount Resize Component"
9. "resize!" messages should stop showing up in the console, but they don't.

I did make some un-orthodox stuff in that hook. But I believe this is a bug since when trying this hook in a production build, it worked fine.

To test out a production build:

(tip: you can skip checking out the repo (steps 1 to 2), as it can be reproduced in [the repo's deployed version](https://use-on.vercel.app/))

1. Check out https://github.com/dutzi/use-on
2. `yarn` then `yarn start:prod`
3. Complete steps 4-9 from above

❤️

**Comment by bvaughn** [comment]
tl;dr The behavior you're pointing out in DEV is intentional. Check out the [React docs on "strict mode"](https://reactjs.org/docs/strict-mode.html) for more background on _why_ React does this.

The `useOn` hook looks like it breaks the rules of React. Calling `.who` during render in your example is a side effect (e.g. it adds event listeners). [Side effects during render will cause problems.](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects)

See related issue #20835 

Going to close this issue since everything looks like it's working as designed on the React side. Hopefully the "strict mode" link I shared will help clear things up! :smile:

**Comment by dutzi** [comment]
Thanks for detailed response! I was guessing it would be some documented behavior I missed :)

**Comment by dutzi** [comment]
@bvaughn, I've made some changes to the hook, I've moved the event listening code inside a `useEffect` that's triggered by a hook that forces a re-render ([`useRerender`](https://github.com/dutzi/use-on/blob/master/src/use-on.ts#L11-L19)). That hook is triggered by calling `who`, `when` or `what`.

That got me a step forward, I was now able to add the listener on mount, and remove it on unmount, _but_, if I [changed `eventName` on the fly](https://github.com/dutzi/use-on/blob/master/src/ListenerComponent.tsx#L15), it didn't remove the old listener (nor did it add the new one).

I went over the code a few times and read more carefully into the docs you've referred to, I hope I didn't miss anything, but it seems pretty side-effects free.

You can see it reproducing (development) [here](https://codesandbox.io/s/github/dutzi/use-on) and working (production) [here](https://use-on.vercel.app/).

Steps to reproduce have changed a bit:

1. Check out https://github.com/dutzi/use-on
2. `yarn` then `yarn start`
3. (Chrome will start up)
4. Open the DevTools' Console
5. Click on "Mount Listener Component"
6. Press any key...
7. (You should see "keydown handler!" printed in the console)
8. Click on "Listen to MouseDown"
9. Click anywhere...
9. You should see "mousedown handler!" printed in the console, but you won't (in development)

As before, the code is [here](https://github.com/dutzi/use-on).
